### ISSUE TITLE
Question about detach in Cycle-GAN?

### ISSUE THREAD

**Issue by kcltfask** [junyanz/pytorch-CycleGAN-and-pix2pix#916]
In script pytorch-CycleGAN-and-pix2pix/models/cycle_gan_model.py, there is a detach() for the fake image before applying netD in backward_D_basic. While the same detach() is not found in others. Is it any special reason? 

    def backward_D_basic(self, netD, real, fake):
        ...
        # Real
        pred_real = netD(real)                                                    <<< no detach()
        loss_D_real = self.criterionGAN(pred_real, True)
        # Fake
        pred_fake = netD(fake.detach())                                     <<< use detach()
        loss_D_fake = self.criterionGAN(pred_fake, False)
        # Combined loss and calculate gradients
        loss_D = (loss_D_real + loss_D_fake) * 0.5
        loss_D.backward()
        return loss_D

    def backward_G(self):
        ...
        # GAN loss D_A(G_A(A))
        self.loss_G_A = self.criterionGAN(self.netD_A(self.fake_B), True)    <<< no detach()
        # GAN loss D_B(G_B(B))
        self.loss_G_B = self.criterionGAN(self.netD_B(self.fake_A), True)    <<< no detach()
        ...

Can both of the them using detach() or NOT using detach() ?
Thank you!

**Comment by junyanz** [comment]
When updating D, we don't need to compute gradients for G. That's why we used `detach` in `backward_D_basic'.

**Comment by kcltfask** [comment]
> When updating D, we don't need to compute gradients for G. That's why we used `detach` in `backward_D_basic`.

Thank you for your reply, that's helpful !!!
And after reading the below article, I finally understand the reason of using detach in above.
[https://blog.csdn.net/Hungryof/article/details/78035332](url)

**Comment by junyanz** [comment]
Glad that you figured it out. I closed this issue.
